### PR TITLE
Refactor subs() of multivariate polynomials for readability and efficiency

### LINKED ISSUE #34591: Refactor subs() of multivariate polynomials for readability and efficiency
<div id="comment:0"></div>

This bug

```
sage: R = Integers(8)
sage: RXY.<X,Y> = R[]
sage: F = X^3*Y
sage: F.subs({X:1})
Y
sage: F.subs({X:2})  # should be 0!
Y
```
is fixed by #34417. However while fixing the bug, it was noticed that the code of `.subs()` is so convoluted that the code is hard to read and inefficient.

So in this ticket, we refactor the code. The modified code is shorter, simpler, and seems slightly faster. 

Moreover the patch also fixes the original bug without #34417 (upstream patch)!

Before the patch:

```sage
sage: R = Integers(8)
sage: RXY.<X,Y> = R[]
sage: F = X^3*Y
sage: F.subs({X:2})
Y
sage: timeit('F.subs({X:2})')
125 loops, best of 3: 3.41 ms per loop
sage: timeit('F.subs({X:2})')
125 loops, best of 3: 3.39 ms per loop
sage: timeit('F.subs({X:2})')
125 loops, best of 3: 3.41 ms per loop
sage: timeit('F.subs({X:2})')
125 loops, best of 3: 3.46 ms per loop
sage: timeit('F.subs({X:2})')
125 loops, best of 3: 3.38 ms per loop
sage: timeit('F.subs({X:1})')
625 loops, best of 3: 8.26 μs per loop
sage: timeit('F.subs({X:1})')
625 loops, best of 3: 8.36 μs per loop
sage: timeit('F.subs({X:1})')
625 loops, best of 3: 8.65 μs per loop
sage: timeit('F.subs({X:1})')
625 loops, best of 3: 8.21 μs per loop
sage: timeit('F.subs({X:1})')
625 loops, best of 3: 8.09 μs per loop
```
After the patch 

```sage
sage: R = Integers(8)
sage: RXY.<X,Y> = R[]
sage: F = X^3*Y
sage: F.subs({X:2})
0
sage: timeit('F.subs({X:2})')
625 loops, best of 3: 8.11 μs per loop
sage: timeit('F.subs({X:2})')
625 loops, best of 3: 8.8 μs per loop
sage: timeit('F.subs({X:2})')
625 loops, best of 3: 8.63 μs per loop
sage: timeit('F.subs({X:2})')
625 loops, best of 3: 8.52 μs per loop
sage: timeit('F.subs({X:2})')
625 loops, best of 3: 8.63 μs per loop
sage: timeit('F.subs({X:1})')
625 loops, best of 3: 8.18 μs per loop
sage: timeit('F.subs({X:1})')
625 loops, best of 3: 7.99 μs per loop
sage: timeit('F.subs({X:1})')
625 loops, best of 3: 8.25 μs per loop
sage: timeit('F.subs({X:1})')
625 loops, best of 3: 8.2 μs per loop
sage: timeit('F.subs({X:1})')
625 loops, best of 3: 8.6 μs per loop
sage: timeit('F.subs({X:1})')
625 loops, best of 3: 8.38 μs per loop
```

CC:  @videlec @yyyyx4

Component: **basic arithmetic**

Author: **Kwankyu Lee**

Branch/Commit: **[u/klee/multivariate_polynomial_variable_substition_has_a_bug](https://github.com/sagemath/sagetrac-mirror/tree/u/klee/multivariate_polynomial_variable_substition_has_a_bug) @ [`4a47cb2`](https://github.com/sagemath/sagetrac-mirror/commit/4a47cb2737817e04f55c478494e3645e8a78272c)**

_Issue created by migration from https://trac.sagemath.org/ticket/34591_

### PR DESCRIPTION
**Pull request by kwankyu**
### 📚 Description

Fixes #34591 

### 📝 Checklist

- [x] I have made sure that the title is self-explanatory and the description concisely explains the PR.
- [x] I have linked an issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/sage/rings/polynomial/multi_polynomial_libsingular.pyx [hunk 1]
```diff
@@ -2077,30 +2077,30 @@ cdef class MPolynomial_libsingular(MPolynomial):
...
+            # give up, evaluate symbolically
             y = parent.base_ring().zero()
-            for (m,c) in self.dict().iteritems():
-                y += c*mul([ x[i]**m[i] for i in m.nonzero_positions()])
+            for m, c in self.dict().iteritems():
+                y += c * mul([x[i]**m[i] for i in m.nonzero_positions()])
             return y
 
-        cdef poly *res    # ownership will be transferred to us in the next line
-        singular_polynomial_call(&res, self._poly, _ring, coerced_x, MPolynomial_libsingular_get_element)
```
Comment (mkoeppe) [inline src/sage/rings/polynomial/multi_polynomial_libsingular.pyx]: I think the comments regarding object ownership may be valuable and should be kept

File: src/sage/rings/polynomial/multi_polynomial_libsingular.pyx [hunk 2]
```diff
@@ -3503,163 +3502,122 @@ cdef class MPolynomial_libsingular(MPolynomial):
...
-                            mi = i
-                            break
-                    if i > _ring.N:
-                        raise TypeError("key does not match")
-                else:
-                    raise TypeError("keys do not match self's parent")
+        # finally change the parent of the result
 
-                g[mi-1] = v
+        res_parent = coercion_model.common_parent(parent._base, *[v for _, v in items])
```
Comment (mkoeppe) [inline src/sage/rings/polynomial/multi_polynomial_libsingular.pyx:3618]: This appears to be more general than the old code that tries to coerce every `v` into `parent._base`.
Is this just incidental, or is there a use case for this greater generality?

File: src/sage/rings/polynomial/multi_polynomial_libsingular.pyx (hunk 1 above)
Comment (kwankyu) [reply src/sage/rings/polynomial/multi_polynomial_libsingular.pyx]: Okay.

File: src/sage/rings/polynomial/multi_polynomial_libsingular.pyx (hunk 2 above)
Comment (kwankyu) [reply src/sage/rings/polynomial/multi_polynomial_libsingular.pyx:3618]: Perhaps there is no difference in this respect. Both first try to coerce every `v` into `parent._base`. If fails, the old code finds the common parent in `return self(*g)` (the last line) via the constructor `__init__` while the new code does the same in the place. However, for some paths (substitutions with constant values), the new code does not make a duplicate call to `singular_polynomial_call` by avoiding going through the constructor. (if I remember correctly...)

Comment (kwankyu) [reply src/sage/rings/polynomial/multi_polynomial_libsingular.pyx:3618]: Perhaps it is the substitution with constant values that *results in zero* that makes the duplicate call.

**Review by mkoeppe — APPROVED** [review]
LGTM

**Comment by vbraun** [comment]
Merge conflict

**Comment by kwankyu** [comment]
Fixed merge conflict with #33373

**Comment by mantepse** [comment]
@kwankyu, sorry to dig up this old PR.  I want to fix #19130, that is, make parents properly depend on the arguments.  Trying to figure out how things work I noticed something I do not understand:

https://github.com/kwankyu/sage/blob/8e0369193b7b83aaf0df8459ec75996ea18a3ba3/src/sage/rings/polynomial/multi_polynomial_libsingular.pyx#L3515-L3536

In the first line of the snippet, `change_ring` is introduced and set to `False`.  I don't see that it would be modified before the last line of the snippet.

Do you remember the intention?

**Comment by kwankyu** [comment]
> In the first line of the snippet, `change_ring` is introduced and set to `False`. I don't see that it would be modified before the last line of the snippet.
> 
> Do you remember the intention?

No. So I read the lines again. 

Line 3536 (`if not change_ring:`) seems unnecessary because `change_ring` is `False` at that moment. 

The main intention of introducing `change_ring` seems this:
```
        We test that we change the ring even if there is nothing to do::

            sage: P = QQ['x,y']
            sage: x = var('x')
            sage: parent(P.zero().subs(x=x))
            Symbolic Ring
```

**Comment by mantepse** [comment]
Well, there must have been *some* intention, because all of this code is new, and was done in an attempt to simplify the code.  Could it be that `change_ring` *should* be modified before the `if`, but wasn't because of an oversight?

(I am trying to figure out what the parent of the result of `subs` should be in principle.)

**Pull request merged**